File: source/CTRW.py
#!/usr/bin/env python
# -*- encoding: utf-8 -*-
# @Time : 2021/12/1 13:27
import numpy as np
from numpy import random
import matplotlib.pyplot as plt
from trajectory import Trajectory
from rnd import stable_rnd, skewed_stable_rnd


class CTRW(Trajectory):
    def __init__(self, t_len, ind_waiting, ind_jump, init_position=0):
        super(Trajectory, self).__init__()
        self._T = t_len
        self._alpha = ind_jump
        self._beta = ind_waiting
        self._x0 = init_position
        self._simulate()

    def _simulate(self):
        if self._beta == 1:
            waiting_time = random.exponential
        else:
            waiting_time = skewed_stable_rnd
        self._t = np.zeros(1)
        self._x = np.array([self._x0])
        total_time = 0
        current_position =  self._x0
        n = 1
        while True:
            tau = waiting_time(self._beta)
            if total_time + tau > self._T:
                self._t = np.append(self._t, self._T)
                self._x = np.append(self._x, current_position)
                self._n = n + 1
                break
            else:
                xi = stable_rnd(self._alpha)
                total_time += tau
                current_position += xi
                self._t = np.append(self._t, total_time)
                self._x = np.append(self._x, current_position)
                n += 1

    def plot(self):
        plt.figure()
        plt.step(self._t, self._x)
        plt.show()


if __name__ == "__main__":
    m1 = CTRW(100, 0.7, 1.5)
    t1, x1 = m1.get()
    fig = plt.figure(1)
    plt.step(t1, x1)
    plt.xlabel("t")
    plt.ylabel("x")
    fig.savefig("ctrw4.eps")

File: MATLAB/CTRW.m
function varargout = CTRW(t_len, alpha, beta, x0)
    if nargin == 3
        x0 = 0;
    end
    
    if alpha == 1
        rand_wait = @(x) exprnd(x); 
    else
        rand_wait = @(x) rand1stable(x);
    end
    
    rand_jump = @(x) rand2stable(x);
        
    total_time = 0;
    current_position = x0;
    n = 1;
    t = zeros;
    x = zeros;
    t(1) = 0;
    x(1) = x0;
    
    while true
        n = n + 1;
        tau = rand_wait(alpha);
        if total_time + tau > t_len
            t(n) = t_len;
            x(n) = current_position;
            break
        else
            total_time = total_time + tau;
            current_position = current_position + rand_jump(beta);
            t(n) = total_time;
            x(n) = current_position;
        end
    end
    
    if nargout == 0
        figure()
        stairs(t, x)
    end
    
    if nargout == 2
        varargout{1} = t;
        varargout{2} = x;
    end
end